### DISCUSSION TITLE
On a binded model validation is done on the casted attributes by default

### DISCUSSION THREAD

**Discussion by barclaymichael** [livewire/livewire#5086] [7. Bugs (archived)]
Context:

Lets say I have a User Model with a property called is_admin that is a boolean. For convenience you cast it as a boolean on the model (otherwise it might be a 0 or 1).

```
class User extends Model
{
   protected $casts = [
        'is_admin' => 'boolean',
    ];
}
```

Now I have a livewire component where a user is binded model. I also have some validation rules, including a boolean requirement.

```
class UserConfigComponent extends Component
{
   public $user;

   protected $rules = [
        'user.is_admin' => ['required','boolean'],
    ];

   public something()
   {
        $this->validate();

        $this->user->save();
   }
}
```

Why is the casting relevant? Let's say I set the $user->is_admin property to something like 300 or "asdf". When is_admin is read, thanks to the casting, it will return true:

```
$user->is_admin = 300; // If I read $user->is_admin it will return "true"
```

Now, when we do validation of EloquentModels in Livewire, the values used are those returned from the ->toArray() function on the model. The logic can be configured on the model, but the default is that $model->toArray() will return the values with their default casts. You can find this logic in the ValidatesInput trait in Livewire, in the unwrapDataForValidation method.

So, in the example above, if we somehow set is_admin to "adafds", when we call `$this->validate()` it will be casted to "true" before validation and validation will pass. This means the user could have sent random characters and we are accepting them. After validation has passed though, when we call $this->user->save() the insert into the database will throw an exception because we will still be trying to insert a string to a boolean column. This results in that validation didn't catch the error early which, I would argue, be one of the goals of the validation step.

My expectation was that validation should be done on the original values entered by the user. This is so we can avoid the casts modifying the input being validated. Maybe a solution looks something like $this->withoutCasting()->validate() that calls an alternative to ->toArray when calling ->validate(). Maybe its the [getAttributes method](https://laravel.com/api/9.x/Illuminate/Database/Eloquent/Concerns/HasAttributes.html#method_getAttributes), but maybe someone has more experience with Eloquent Models to offer a different opinion.

Hope this makes sense, and thanks for reading.

**Comment by Martleby** [comment]
Wow, that's a bit of a gotcha!

I agree with you - instinctively I think the validation should happen against the model's attribute bag values by default. This seems to me the safest thing to do (and also it'd be easier because there'd be none of that reflection api stuff). As per your example it's clear that Eloquent uses the attributes rather than casted values to save to the DB.

That said, I wonder if bools are one of those funny edge cases (thanks to what PHP considers to be truthy and falsy values) and in the majority of non-bool cases this is never an issue? Might be worth experimenting with that to see if this could cause problems with anything else.

Anyway, thanks for the pointer on this one, I'm going to keep it under my hat. Tomorrow I'm going to review my code to see where this might be an issue. I guess the interim answer is to hook into an updatingUserIsAdmin method and calling the validation before it gets set on the model.

Cheers,
Martin.
